transports/noise/src/io/framed.rs

```diff
@@ -98,278 +90,141 @@ impl<T> NoiseFramed<T, snow::HandshakeState> {
     /// an error is returned. Similarly if the remote's static DH key, if
     /// present, cannot be parsed.
     pub(crate) fn into_transport(self) -> Result<(PublicKey, Output<T>), Error> {
-        let dh_remote_pubkey = self.session.get_remote_static().ok_or_else(|| {
+        let dh_remote_pubkey = self.io.codec().session.get_remote_static().ok_or_else(|| {
             Error::Io(io::Error::new(
                 io::ErrorKind::Other,
                 "expect key to always be present at end of XX session",
             ))
         })?;
 
         let dh_remote_pubkey = PublicKey::from_slice(dh_remote_pubkey)?;
+        let framed_parts = self.io.into_parts();
 
         let io = NoiseFramed {
-            session: self.session.into_transport_mode()?,
-            io: self.io,
-            read_state: ReadState::Ready,
-            write_state: WriteState::Ready,
-            read_buffer: self.read_buffer,
-            write_buffer: self.write_buffer,
-            decrypt_buffer: self.decrypt_buffer,
+            io: asynchronous_codec::Framed::new(
+                framed_parts.io,
+                Codec {
+                    session: framed_parts.codec.session.into_transport_mode()?,
+                    write_buffer: Vec::new(),
+                    decrypt_buffer: BytesMut::new(),
+                },
+            ),
         };
 
         Ok((dh_remote_pubkey, Output::new(io)))
     }
 }
 
-/// The states for reading Noise protocol frames.
-#[derive(Debug)]
-enum ReadState {
-    /// Ready to read another frame.
-    Ready,
-    /// Reading frame length.
-    ReadLen { buf: [u8; 2], off: usize },
-    /// Reading frame data.
-    ReadData { len: usize, off: usize },
-    /// EOF has been reached (terminal state).
-    ///
-    /// The associated result signals if the EOF was unexpected or not.
-    Eof(Result<(), ()>),
-    /// A decryption error occurred (terminal state).
-    DecErr,
+/// Codec for writing and reading length-delimited noise messages.
+pub(crate) struct Codec<S> {
+    write_buffer: Vec<u8>,
+    decrypt_buffer: BytesMut,
+    session: S,
 }
 
-/// The states for writing Noise protocol frames.
-#[derive(Debug)]
-enum WriteState {
-    /// Ready to write another frame.
-    Ready,
-    /// Writing the frame length.
-    WriteLen {
-        len: usize,
-        buf: [u8; 2],
-        off: usize,
-    },
-    /// Writing the frame data.
-    WriteData { len: usize, off: usize },
-    /// EOF has been reached unexpectedly (terminal state).
-    Eof,
-    /// An encryption error occurred (terminal state).
-    EncErr,
+impl<S> asynchronous_codec::Encoder for Codec<S>
+where
+    S: SessionState,
```

Comment:
This whole `SessionState` concept differentiates between the initial handshake and the "working mode" afterwards. I wonder if it would make sense to instead of having this `impl` be generic over `S`, we instead to the following:

```rust
impl asynchronous_codec::Encoder for Codec<noise::HandshakeState> {
	type Error = io::Error;
	type Item = proto::NoiseHandshakePayload // <-- Note this
}

impl asynchronous_codec::Encoder for Codec<noise::TransportState> {
	type Error = io::Error;
	type Item = Vec<u8>;
}
```

We'd have to work out a way of reducing some of the code duplication between the two. We could either keep the trait and add some utility functions for reading and writing or we remove the trait and try to de-duplicate the other bits as functions like:

```rust
fn write_encrypted_bytes(&mut self, dst: &mut BytesMut, n_bytes: usize) {
	let prefix = u16::to_be_bytes(n_bytes as u16);
	
	self.write_buffer.truncate(n_bytes);

	dst.put(&prefix[..]);
	dst.put(&self.write_buffer[..]);
}
```

I'd be curious how that turns out. It might allow us to remove / simplify some layers further up because we can directly hand a `proto::NoiseHandshakePayload` to the codec.

What do you think?

Reply:
I think it can be done without much duplication at all(but maybe I'm missing something). Side note that the `Item` for the handshake impl would have to be some kind of enum to allow `send_empty` and `recv_empty`.
```rust
enum Payload {
    Empty,
    NonEmpty(proto::NoiseHandshakePayload)
}
```

Reply:
I think `proto::NoiseHandshakePayload::default()` should give you an empty `Vec<u8>` when serialized so we might not need the enum.